File: python/packages/extractor/src/insightgraph_extractor/claim.py
from __future__ import annotations

import asyncio
import json
import logging
from typing import Any
from uuid import UUID

import litellm

from insightgraph_core.ir.extraction import ExtractedClaim
from insightgraph_core.ir.models import Block
from insightgraph_core.types import ClaimType
from insightgraph_extractor.base import BaseExtractor
from insightgraph_extractor.prompts.claim import (
    CLAIM_SYSTEM_PROMPT,
    format_claim_prompt,
)

logger = logging.getLogger(__name__)

_BATCH_SIZE = 5
_MAX_CONCURRENCY = 4

# Valid claim type values for fast lookup.
_VALID_CLAIM_TYPES = {c.value for c in ClaimType}


def _parse_claim_type(raw: str) -> ClaimType:
    """Safely parse a claim type string, defaulting to FACTUAL."""
    normalized = raw.strip().upper()
    if normalized in _VALID_CLAIM_TYPES:
        return ClaimType(normalized)
    return ClaimType.FACTUAL


def _parse_claims(raw_json: str, block_ids: list[UUID]) -> list[ExtractedClaim]:
    """Parse LLM JSON response into ExtractedClaim objects."""
    try:
        data: dict[str, Any] = json.loads(raw_json)
    except (json.JSONDecodeError, TypeError):
        logger.warning("Failed to parse claim extraction JSON: %s", raw_json[:200])
        return []

    claims_raw = data.get("claims")
    if not isinstance(claims_raw, list):
        logger.warning("Expected 'claims' list in response, got %s", type(claims_raw))
        return []

    default_block_id = block_ids[0] if block_ids else None
    results: list[ExtractedClaim] = []

    for item in claims_raw:
        if not isinstance(item, dict):
            continue

        text = (item.get("text") or "").strip()
        if not text:
            continue

        if default_block_id is None:
            continue

        claim_type = _parse_claim_type(item.get("type", "FACTUAL"))

        entities_mentioned_raw = item.get("entities_mentioned", [])
        if not isinstance(entities_mentioned_raw, list):
            entities_mentioned_raw = []
        entities_mentioned = [
            str(e).strip() for e in entities_mentioned_raw if isinstance(e, str) and str(e).strip()
        ]

        source_text = (item.get("source_text") or "").strip()

        results.append(
            ExtractedClaim(
                text=text,
                type=claim_type,
                entities_mentioned=entities_mentioned,
                source_block_id=default_block_id,
                source_text=source_text,
            )
        )

    return results


class ClaimExtractor(BaseExtractor):
    """Extracts claims and assertions from document blocks using an LLM."""

    def __init__(
        self,
        model: str = "gpt-4o-mini",
        api_key: str = "",
        batch_size: int = _BATCH_SIZE,
        max_concurrency: int = _MAX_CONCURRENCY,
    ) -> None:
        self.model = model
        self.api_key = api_key
        self.batch_size = batch_size
        self._semaphore = asyncio.Semaphore(max_concurrency)

    # -- public API -----------------------------------------------------------

    async def extract(
        self, blocks: list[Block], context: dict | None = None
    ) -> list[ExtractedClaim]:
        """Extract claims from *blocks*."""
        if not blocks:
            return []

        ctx = context or {}
        doc_title = ctx.get("title") or "Unknown"

        batches = self._make_batches(blocks)
        tasks = [self._extract_batch(batch, doc_title=doc_title) for batch in batches]
        batch_results = await asyncio.gather(*tasks, return_exceptions=True)

        all_claims: list[ExtractedClaim] = []
        for result in batch_results:
            if isinstance(result, BaseException):
                logger.error("Claim extraction batch failed: %s", result)
                continue
            all_claims.extend(result)

        return all_claims

    # -- internals ------------------------------------------------------------

    def _make_batches(self, blocks: list[Block]) -> list[list[Block]]:
        return [blocks[i : i + self.batch_size] for i in range(0, len(blocks), self.batch_size)]

    async def _extract_batch(self, batch: list[Block], doc_title: str) -> list[ExtractedClaim]:
        combined_text = "\n\n".join(block.content for block in batch)
        block_ids = [block.id for block in batch]

        section_title = "Unknown"
        for block in batch:
            if block.level is not None:
                section_title = block.content
                break

        user_prompt = format_claim_prompt(
            text=combined_text,
            doc_title=doc_title,
            section_title=section_title,
        )

        async with self._semaphore:
            try:
                response = await litellm.acompletion(
                    model=self.model,
                    api_key=self.api_key or None,
                    messages=[
                        {"role": "system", "content": CLAIM_SYSTEM_PROMPT},
                        {"role": "user", "content": user_prompt},
                    ],
                    response_format={"type": "json_object"},
                    temperature=0.0,
                )
            except Exception:
                logger.exception("LLM call failed for claim extraction batch")
                return []

        raw_content = response.choices[0].message.content or ""
        claims = _parse_claims(raw_content, block_ids)

        # Match source_text to the correct block.
        claims = self._match_source_blocks(claims, batch)
        return claims

    @staticmethod
    def _match_source_blocks(
        claims: list[ExtractedClaim], blocks: list[Block]
    ) -> list[ExtractedClaim]:
        """Best-effort assignment of each claim's source_block_id."""
        updated: list[ExtractedClaim] = []
        for claim in claims:
            if claim.source_text:
                for block in blocks:
                    if claim.source_text in block.content:
                        claim = claim.model_copy(update={"source_block_id": block.id})
                        break
            updated.append(claim)
        return updated

File: packages/extractor/src/claim.ts
import type { Block, ExtractedClaim, ClaimType } from "@insightgraph/core";
import { createLLMClient, chatJSON, CLAIM_TYPES, safeParseLlmJson, isRecord } from "@insightgraph/core";
import type { LLMClient } from "@insightgraph/core";
import { CLAIM_SYSTEM_PROMPT, formatClaimPrompt } from "./prompts/claim";
import { createLimiter } from "./concurrency";
import {
  DEFAULT_BATCH_SIZE,
  DEFAULT_MAX_CONCURRENCY,
  type ExtractorOptions,
} from "./options";

const VALID_CLAIM_TYPES = new Set<string>(CLAIM_TYPES);

export class ClaimExtractor {
  private client: LLMClient;
  private model: string;
  private batchSize: number;
  private maxConcurrency: number;

  constructor(
    model: string,
    apiKey: string,
    baseUrl = "",
    options: ExtractorOptions = {},
  ) {
    this.client = createLLMClient(apiKey, baseUrl || undefined);
    this.model = model;
    this.batchSize = options.batchSize ?? DEFAULT_BATCH_SIZE;
    this.maxConcurrency = options.maxConcurrency ?? DEFAULT_MAX_CONCURRENCY;
  }

  async extract(blocks: Block[], context?: { title?: string }): Promise<ExtractedClaim[]> {
    const limit = createLimiter(this.maxConcurrency);
    const docTitle = context?.title ?? "Unknown";

    const batches = makeBatches(blocks, this.batchSize);
    const tasks = batches.map((batch) =>
      limit(() => this.extractBatch(batch, docTitle)),
    );
    const results = await Promise.allSettled(tasks);

    const all: ExtractedClaim[] = [];
    for (const r of results) {
      if (r.status === "fulfilled") all.push(...r.value);
    }
    return all;
  }

  private async extractBatch(batch: Block[], docTitle: string): Promise<ExtractedClaim[]> {
    const combinedText = batch.map((b) => b.content).join("\n\n");
    const blockIds = batch.map((b) => b.id);

    let sectionTitle = "Unknown";
    for (const block of batch) {
      if (block.level != null) { sectionTitle = block.content; break; }
    }

    try {
      const raw = await chatJSON(this.client, this.model, [
        { role: "system", content: CLAIM_SYSTEM_PROMPT },
        { role: "user", content: formatClaimPrompt(combinedText, docTitle, sectionTitle) },
      ]);
      return parseClaims(raw, blockIds);
    } catch (err) {
      console.warn(`[claim-extractor] LLM call failed: ${(err as Error).message}`);
      return [];
    }
  }
}

function parseClaimType(raw: string): ClaimType {
  const upper = raw.toUpperCase();
  return VALID_CLAIM_TYPES.has(upper) ? (upper as ClaimType) : "FACTUAL";
}

function parseClaims(rawJson: string, blockIds: string[]): ExtractedClaim[] {
  if (blockIds.length === 0) return [];
  const data = safeParseLlmJson<{ claims?: Array<Record<string, unknown>> }>(
    rawJson,
    { context: "claim-extractor", validate: isRecord },
  );
  if (!data) return [];

  const claims: ExtractedClaim[] = [];
  for (const c of data.claims ?? []) {
    if (!c.text) continue;
    claims.push({
      text: c.text as string,
      type: parseClaimType((c.type as string | undefined) ?? "FACTUAL"),
      entitiesMentioned: (c.entities_mentioned as string[]) ?? [],
      confidence: typeof c.confidence === "number" ? c.confidence : 1.0,
      sourceBlockId: blockIds[0],
      sourceText: (c.source_text as string | undefined) ?? "",
    });
  }
  return claims;
}

function makeBatches<T>(items: T[], size: number): T[][] {
  const batches: T[][] = [];
  for (let i = 0; i < items.length; i += size) {
    batches.push(items.slice(i, i + size));
  }
  return batches;
}
